pkg/storage/iterators.go

```diff
@@ -102,38 +102,41 @@ func (u *uniqueObjectIterator) Stop() {
 }
 
 type combinedIterator[T any] struct {
-	iter1, iter2 Iterator[T]
+	iters []Iterator[T]
 }
 
 func (c *combinedIterator[T]) Next() (T, error) {
-	val, err := c.iter1.Next()
-	if err != nil {
-		if !errors.Is(err, ErrIteratorDone) {
-			return val, err
-		}
-	} else {
-		return val, nil
-	}
-
-	val, err = c.iter2.Next()
-	if err != nil {
-		if !errors.Is(err, ErrIteratorDone) {
-			return val, err
+	for i := 0; i < len(c.iters); i++ {
+		if c.iters[i] != nil {
+			val, err := c.iters[i].Next()
+			if err != nil {
+				if !errors.Is(err, ErrIteratorDone) {
+					return val, err
+				} else {
+					c.iters[i] = nil // end of this iterator
+				}
+			} else {
```

Comment:
You can remove excessive  indentation and make this code more easily readable with:
```

for i, iter := range c.iters {
    if iter == nil {
        continue
    }

    val, err := iter.Next()
    if err != nil {
        if !errors.Is(err, ErrIteratorDone) {
            return val, err
        }

       c.iters[i] = nil
       continue
    }

    return val, nil
}
```